cannon/mipsevm/tests/helpers.go

```diff
@@ -50,31 +50,50 @@ func multiThreadElfVmFactory(t require.TestingT, elfFile string, po mipsevm.Prei
 	return fpvm
 }
 
+type ThreadProofEncoder func(po mipsevm.PreimageOracle, stdOut, stdErr io.Writer, log log.Logger, opts ...testutil.StateOption) []byte
```

Comment:
I think we could simplify a bit and make this new util more generally useful if we just accept a state directly and generate the whole proof: 
```suggestion
type ProofGenerator func(t require.TestingT, state FPVMState, memoryProofAddresses ...uint32) []byte
```

For example, the multithreaded one could look something like:
```
func multiThreadedProofGenerator(t require.TestingT, state FPVMState, memoryProofAddresses ...uint32) []byte {
    mtState, ok := state.(*multithreaded.State)
    if !ok {
        require.Fail(t, "Failed to cast FPVMState to multithreaded State type")
    }

    proofData := mtState.EncodeThreadProof()
    insnProof := mtState.GetMemory().MerkleProof(mtState.GetPC())
    proofData = append(proofData, insnProof[:]...)

    for _, addr := range memoryProofAddresses {
        memProof := mtState.GetMemory().MerkleProof(addr)
        proofData = append(proofData, memProof)
    }

    return proofData
}
```

Reply:
nice!